### PR TITLE
[RLlib] Issue 22179: Apex (DQN and DDPG) multi-GPU support.

### LINKED ISSUE #22179: [Bug] [RLLIB] APEX-DDPG crashes if using multiple gpus
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

I tried out the torch implementation of APEX-DDPG with multiple gpus but it crashes.

### Versions / Dependencies

ray 1.10.0

### Reproduction script

I tried running the following config (same as the one in tuned examples but with torch and 2 gpus) using `rllib train -f xxx.yaml` : 

```yaml
mountaincarcontinuous-apex-ddpg:
    env: MountainCarContinuous-v0
    run: APEX_DDPG
    stop:
        episode_reward_mean: 90
    config:
        # Works for both torch and tf.
        framework: torch
        clip_rewards: False
        num_gpus: 2
        num_workers: 16
        exploration_config:
            ou_base_scale: 1.0
        n_step: 3
        target_network_update_freq: 50000
        tau: 1.0
        evaluation_interval: 5
        evaluation_num_episodes: 10
```

### Anything else

The error below is systematic. I can run the Impala multi-gpu config just fine (https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/impala/atari-impala-multi-gpu.yaml), so that shouldn't be hardware related.

Here is the error : 

```python
2022-02-07 17:10:29,203 ERROR trial_runner.py:927 -- Trial APEX_DDPG_MountainCarContinuous-v0_7075d_00000: Error processing event.
Traceback (most recent call last):
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 893, in _process_trial
    results = self.trial_executor.fetch_result(trial)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py", line 707, in fetch_result
    result = ray.get(trial_future[0], timeout=DEFAULT_GET_TIMEOUT)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/worker.py", line 1733, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(RuntimeError): ray::ApexDDPGTrainer.train() (pid=941158, ip=192.168.251.123, repr=ApexDDPGTrainer)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/tune/trainable.py", line 315, in train
    result = self.step()
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 982, in step
    raise e
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 963, in step
    step_attempt_results = self.step_attempt()
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 1042, in step_attempt
    step_results = self._exec_plan_or_training_iteration_fn()
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 1962, in _exec_plan_or_training_iteration_fn
    results = next(self.train_exec_impl)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 756, in __next__
    return next(self.built_iterator)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 843, in apply_filter
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 1075, in build_union
    item = next(it)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 756, in __next__
    return next(self.built_iterator)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/util/iter.py", line 783, in apply_foreach
    for item in it:
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/execution/concurrency_ops.py", line 135, in base_iterator
    raise RuntimeError("Dequeue `check()` returned False! "
RuntimeError: Dequeue `check()` returned False! Exiting with Exception from Dequeue iterator.
== Status ==
Current time: 2022-02-07 17:10:29 (running for 00:00:12.37)
Memory usage on this node: 43.3/125.7 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/64 CPUs, 0/2 GPUs, 0.0/56.88 GiB heap, 0.0/28.37 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/delfox/ray_results/mountaincarcontinuous-apex-ddpg
Number of trials: 1/1 (1 ERROR)
+------------------------------------------------+----------+------------------------+
| Trial name                                     | status   | loc                    |
|------------------------------------------------+----------+------------------------|
| APEX_DDPG_MountainCarContinuous-v0_7075d_00000 | ERROR    | 192.168.251.123:941158 |
+------------------------------------------------+----------+------------------------+
Number of errored trials: 1
+------------------------------------------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| Trial name                                     |   # failures | error file                                                                                                                              |
|------------------------------------------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------|
| APEX_DDPG_MountainCarContinuous-v0_7075d_00000 |            1 | /home/delfox/ray_results/mountaincarcontinuous-apex-ddpg/APEX_DDPG_MountainCarContinuous-v0_7075d_00000_0_2022-02-07_17-10-16/error.txt |
+------------------------------------------------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------+

Traceback (most recent call last):
  File "/home/delfox/.opt/miniconda3/envs/swarm/bin/rllib", line 8, in <module>
    sys.exit(cli())
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/scripts.py", line 36, in cli
    train.run(options, train_parser)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/rllib/train.py", line 267, in run
    concurrent=True)
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/tune/tune.py", line 739, in run_experiments
    callbacks=callbacks).trials
  File "/home/delfox/.opt/miniconda3/envs/swarm/lib/python3.7/site-packages/ray/tune/tune.py", line 630, in run
    raise TuneError("Trials did not complete", incomplete_trials)
ray.tune.error.TuneError: ('Trials did not complete', [APEX_DDPG_MountainCarContinuous-v0_7075d_00000])
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by sven1977**
Issue 22179: Apex (DQN and DDPG) multi-GPU support.

## Why are these changes needed?

Issue #22179

## Related issue number

Closes #22179 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/dqn/learner_thread.py [hunk 1]
```diff
@@ -54,22 +84,115 @@ def step(self):
...
+            self.loader_thread = _MultiGPULoaderThread(self, share_stats=(i == 0))
+            self.loader_thread.start()
+
+    @override(APEXLearnerThread)
+    def step(self) -> None:
+        assert self.loader_thread.is_alive()
+        with self.load_wait_timer:
+            replay_actor, buffer_idx, batch_indexes = self.ready_tower_stacks.get()
+
+        get_num_samples_loaded_into_buffer = 0
```
Comment (gjoliver) [inline rllib/agents/dqn/learner_thread.py:173]: this variable probably should be named num_samples_loaded_into_buffer

File: rllib/agents/dqn/learner_thread.py [hunk 2]
```diff
@@ -54,22 +84,115 @@ def step(self):
...
+                data into GPU memory in parallel.
+        """
+        super().__init__(local_worker=local_worker)
+
+        self.train_batch_size = train_batch_size
+
+        # TODO: (sven) Allow multi-GPU to work for multi-agent as well.
+        self.policy = self.local_worker.policy_map[DEFAULT_POLICY_ID]
+        self.policy_map = self.local_worker.policy_map
+        self.devices = self.policy.devices
```
Comment (gjoliver) [inline rllib/agents/dqn/learner_thread.py:148]: curious, where do we verify that there are that many physical GPU devices for batch loading and learning?
does it happen in policy?

**Review by gjoliver — COMMENTED** [review]
cool, a couple of very quick questions.

File: rllib/agents/dqn/learner_thread.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/dqn/learner_thread.py:148]: Yes, in the Policy c'tor. There, we check the config as well as the ray assigned devices and then check for their validity using tf and torch tools.

**Comment by sven1977** [comment]
ping

File: rllib/agents/dqn/learner_thread.py [hunk 3]
```diff
@@ -54,22 +84,115 @@ def step(self):
...
+            )
+
+            self.learner_info = learner_info_builder.finalize()
+
+            td_error = self.learner_info["default_policy"]["td_error"]
+            prio_dict = {"default_policy": (batch_indexes, td_error)}
+
+        self.idle_tower_stacks.put(buffer_idx)
+        self.outqueue.put((replay_actor, prio_dict, get_num_samples_loaded_into_buffer))
+        self.learner_queue_size.push(self.inqueue.qsize())
```
Comment (gjoliver) [inline rllib/agents/dqn/learner_thread.py:198]: a random question, do we need to care about multi-threading related issues with these learner threads? not sure if there are more than 1 such threads actually.

**Review by gjoliver — COMMENTED** [review]
very sorry about the delay.
just a few quick questions now.

**Pull request closed without merging**